### DISCUSSION TITLE
Conditional styling of dynamic checkboxes

### DISCUSSION THREAD

**Discussion by alinawinter** [neuefische/questions#569] [Web] [answered]
Hey,

I'm currently working on my form component (FormPreselection) and thought this morning I'd quickly do some styling for my checkboxes (which I dynamically map inside the form from a database). 

3 hours later, I realize it's not as easy as I had thought. All the styling on the checkboxes worked quite well. What doesn't work is changing the background color of the CheckBoxContainer, when the checkboxes are checked. Because the checkboxes are developed dynamically, it changes the background color for all CheckBoxContainers, if I add a ternary operator like "isChecked ? X : Y" to it. If I click on an element again to unclick the checkbox, it also doesn't change the background color. Also when the reset button is clicked, this doesn't have an effect on the background color. 

What I tried this far (but deleted again, because it didn't work):
- creating a state isChecked, that I update in the onChange-function of the input field and hand as a prop to the CheckBoxContainer (leads to the issue that all Containers turn background-color by clicking one)
- mapping through the selectedFoodCategories-Array that I create based on clicked checkboxes - however, this array only fully updates when the submit button is clicked, it is always "one click behind", also here if I unclick a checkbox, the background color does not change. 
- searching the Internet and asking Chat GPT

I committed my current status and pushed it. You can view it here: https://github.com/alinawinter/capstone-project/pull/14
This is the deployed version: https://capstone-project-ayztwz2rx-alinawinter.vercel.app/

I really hope someone has a solution! Thanks alot!

**Comment by PReichetanz** [comment, marked as answer]
Hi @alinawinter,

I've found a way (might not be the best one, but it works). The main point is that you need to save somewhere which of the checkboxes is checked and which is not. I've done this in the `selectedFoodCategories` state.

## Step 1 Toggle an `isChecked` key in the state
- Refactor the function `handleAddSelectedFoodCategories` to add the selected food to the state if it's not in there and add an `isChecked` key set to `true`.
- Add an else case to remove the selectedFood from the array if the function is called and the food is already in `selectedFoodCategories` (= was checked before).
- Rename the function because it now handles both cases, add and delete
- Remove the `handleDeleteSelectedFoodCategories` function because it is not necessary anymore.

```jsx
// _app.js
function handleSelectedFoodCategories(selection) {
    const assignedFoodCategoryFromDB = foodCategories.find(
      (foodCategory) => foodCategory.name === selection
    );

    if (
      !selectedFoodCategories.find(
        (foodCategory) => foodCategory.name === assignedFoodCategoryFromDB.name
      )
    ) {
      const newAssignedFoodCategory = {
        ...assignedFoodCategoryFromDB,
        consumedQuantity: 0,
        isChecked: true,
      };
      setSelectedFoodCategories((prevSelected) => [
        ...prevSelected,
        newAssignedFoodCategory,
      ]);
    } else {
      const filteredArray = selectedFoodCategories.filter(
        (foodCategory) => foodCategory.name !== selection
      );
      setSelectedFoodCategories([...filteredArray]);
    }
```

## Step 2 Adapt FormPreselectionFood component
- Refactor the `handleCheckBoxChange` function according to the above changes: remove the `if` statement and the `isChecked` variable.

```jsx
function handleCheckBoxChange(event) {
    const checkBoxValue = event.target.value;
    handleAddSelectedFoodCategories(checkBoxValue);
  }
```

- Evaluate whether the current checkbox is checked and pass the boolean result as a prop:

```jsx
<CheckBoxContainer
            key={id}
            isChecked={selectedFoodCategories.find(
              (foodCategory) => foodCategory.name === name
            )}
          >
```

- Style the background according to the `isChecked` prop passed to the CheckBoxContainer above:

```jsx
const CheckBoxContainer = styled.div`
   ...
  background-color: ${({ isChecked }) =>
    isChecked ? 'crimson' : 'rgba(242, 204, 143, 0.4)'};
```

I hope this helps, please let me know if you have any further issues!

**Reply by alinawinter** [reply]
Thanks you a hundred times!! It worked. 
The only thing that did not work yet was the reset button - I assume, because it handled the reset with "event.target.form.reset()". I solved that by on click emptying my array (and thereby the isChecked-values) using setSelectedFoodCategories([]).
(I transformed the reset button into a toggle button with "Select All" and "Reset")

.../FormPreselection/index.js
`  function handleMarkAll(event) {
    event.preventDefault();
    if (allIsMarked === true) {
      setSelectedFoodCategories([]);
      setAllIsMarked(false);
      setButtonText("Select all");
    }
    if (allIsMarked === false) {
      const checkedFoodCategories = foodCategories.map((foodCategory) => ({
        ...foodCategory,
        consumedQuantity: 0,
        isChecked: true,
      }));
      setSelectedFoodCategories([...checkedFoodCategories]);
      setAllIsMarked(true);
      setButtonText("Reset");
    }
  }`

`<BasicButton onClick={handleMarkAll}>{buttonText}</BasicButton>
`
